src/tbb/threading_control.cpp

```diff
@@ -0,0 +1,397 @@
+    // This check relies on the fact that for shared RML default_concurrency == max_concurrency
+    if (!governor::UsePrivateRML && td->my_server->default_concurrency() < workers_soft_limit) {
+        runtime_warning("RML might limit the number of workers to %u while %u is requested.\n",
+            td->my_server->default_concurrency(), workers_soft_limit);
+    }
+
+    return td;
+}
+
+threading_control_impl::threading_control_impl(threading_control* tc) {
+    unsigned workers_soft_limit{}, workers_hard_limit{};
+    std::tie(workers_soft_limit, workers_hard_limit) = calculate_workers_limits();
+
+    my_permit_manager = make_permit_manager(workers_soft_limit);
+    my_thread_dispatcher = make_thread_dispatcher(*tc, workers_soft_limit, workers_hard_limit);
+    my_thread_request_serializer =
+        d1::make_cache_aligned_unique<thread_request_serializer_proxy>(*my_thread_dispatcher, workers_soft_limit);
+    my_permit_manager->set_thread_request_observer(*my_thread_request_serializer);
+
+    my_cancellation_disseminator = d1::make_cache_aligned_unique<cancellation_disseminator>();
+    my_waiting_threads_monitor = d1::make_cache_aligned_unique<thread_control_monitor>();
+}
+
+void threading_control_impl::release(bool blocking_terminate) {
+    my_thread_dispatcher->my_join_workers = blocking_terminate;
+    my_thread_dispatcher->my_server->request_close_connection();
+}
+
+void threading_control_impl::set_active_num_workers(unsigned soft_limit) {
+    __TBB_ASSERT(soft_limit <= my_thread_dispatcher->my_num_workers_hard_limit, nullptr);
+    my_thread_request_serializer->set_active_num_workers(soft_limit);
+    my_permit_manager->set_active_num_workers(soft_limit);
+}
+
+threading_control_impl::threading_control_client threading_control_impl::create_client(arena& a) {
+    pm_client* pm_client = my_permit_manager->create_client(a);
+    thread_dispatcher_client* td_client = my_thread_dispatcher->create_client(a);
+
+    return {pm_client, td_client};
+}
+
+threading_control_impl::client_deleter threading_control_impl::prepare_client_destruction(threading_control_impl::threading_control_client client) {
+    thread_dispatcher_client* td_client = client.second;
+    return {td_client->get_aba_epoch(), td_client->priority_level(), td_client, static_cast<pm_client*>(client.first)};
+}
+
+bool threading_control_impl::try_destroy_client(threading_control_impl::client_deleter deleter) {
+    if (my_thread_dispatcher->try_unregister_client(deleter.my_td_client, deleter.aba_epoch, deleter.priority_level)) {
+        my_permit_manager->destroy_client(*deleter.my_pm_client);
+        return true;
+    }
+    return false;
+}
+
+void threading_control_impl::publish_client(threading_control_impl::threading_control_client client) {
+    my_permit_manager->register_client(static_cast<pm_client*>(client.first));
+    my_thread_dispatcher->register_client(client.second);
```

Comment:
I would recommend to avoid using pairs since they are hard to read. One would need to jump back and forth in the code in order to understand what this `client.second` means. As for me, the whole lines says to thread dispatcher "register a second client for me", which is, obviously, not correct.
But how to change it so that it is read correctly?